mllib/src/main/scala/org/apache/spark/ml/feature/Imputer.scala

```diff
@@ -0,0 +1,300 @@
+  extends Estimator[ImputerModel] with ImputerParams with DefaultParamsWritable {
+
+  @Since("2.0.0")
+  def this() = this(Identifiable.randomUID("imputer"))
+
+  /** @group setParam */
+  def setInputCol(value: String): this.type = set(inputCol, value)
+
+  /** @group setParam */
+  def setOutputCol(value: String): this.type = set(outputCol, value)
+
+  /**
+   * Imputation strategy. Available options are "mean", "median" and "most".
+   * @group setParam
+   */
+  def setStrategy(value: String): this.type = set(strategy, value)
+
+  /** @group setParam */
+  def setMissingValue(value: Double): this.type = set(missingValue, value)
+
+  setDefault(strategy -> "mean", missingValue -> Double.NaN)
+
+  override def fit(dataset: DataFrame): ImputerModel = {
+    val alternate = dataset.select($(inputCol)).schema.fields(0).dataType match {
+      case DoubleType =>
+        val doubleRDD = dataset.select($(inputCol)).rdd.map(_.getDouble(0))
+        Vectors.dense(getColStatistics(doubleRDD))
+      case _: VectorUDT =>
+        val filteredDF = dataset.where(s"${$(inputCol)} IS NOT NULL").select($(inputCol))
+        val vectorRDD = filteredDF.rdd.map(_.getAs[Vector](0)).cache()
+        val vl = dataset.first().getAs[Vector]($(inputCol)).size
+        $(strategy) match {
+          case "mean" =>
+            val summary = vectorRDD.treeAggregate((new Array[Double](vl), new Array[Int](vl)))(
+              (prev, data) => (prev, data) match { case ((mean, count), data) =>
+                  var i = 0
+                  while (i < mean.length) {
+                    if (data(i) != 0 && !data(i).isNaN) {
+                      count(i) += 1
+                      mean(i) = mean(i) + (data(i) - mean(i)) / count(i)
+                    }
+                    i += 1
+                  }
+                  (mean, count)
+              }, (aggregator1, aggregator2) => (aggregator1, aggregator2) match {
+                case ((mean1, c1), (mean2, c2)) =>
+                  (0 until mean1.length).foreach{ i =>
+                    mean1(i) = mean1(i) + (mean2(i) - mean1(i)) * c2(i) / (c1(i) + c2(i))
+                    c1(i) += c2(i)
+                  }
+                  (mean1, c1)
+              })
+            Vectors.dense(summary._1)
+          case _ =>
+            val statisticsArray = new Array[Double](vl)
+            (0 until vl).foreach(i => {
+              statisticsArray(i) = getColStatistics(vectorRDD.map(v => v(i)))
+            })
+            Vectors.dense(statisticsArray)
+        }
+    }
+    copyValues(new ImputerModel(uid, alternate).setParent(this))
+  }
+
+  /** Extract the statistics info from a Double column according to the strategy */
+  private def getColStatistics(data: RDD[Double]): Double = {
+    val filteredRDD = data.filter(!isMissingValue(_))
+    val colStatistics = $(strategy) match {
+      case "mean" => filteredRDD.mean()
+      case "median" => filteredRDD.sortBy(d => d).zipWithIndex()
```

Comment:
We did not open up the multiple columns API because approxQuantile was still experimental.  Originally, we wanted to provide approx quantiles via dataframe.describe, but decided it would make describe() too slow.  It's provided as a stat function for now, until we either speed it up or decide to keep it as a stat function.  I'm OK with opening up the multiple-column API.